source/dubregistry/internal/utils.d

```diff
@@ -13,3 +28,78 @@ string black(string url)
 @safe {
 	return black(URL(url)).toString();
 }
+
+/**
+ * Params:
+ *   file = the file to convert
+ * Returns: the PNG stream of the icon or empty on failure
+ * Throws: Exception if input is invalid format, invalid dimension or times out
+ */
+bdata_t generateLogo(NativePath file) @safe
+{
+	static assert (isWeaklyIsolated!(typeof(&generateLogoUnsafe)));
+	static assert (isWeaklyIsolated!NativePath);
+	static assert (isWeaklyIsolated!LogoGenerateResponse);
+	auto res = (() @trusted => async(&generateLogoUnsafe, file).getResult())();
+	if (res.error.length)
+		throw new Exception("Failed to generate logo: " ~ res.error);
+	return res.data;
+}
+
+private struct LogoGenerateResponse
+{
+	bdata_t data;
+	string error;
+}
+
+// need to return * here because stack returned values get destroyed for some reason...
+private LogoGenerateResponse* generateLogoUnsafe(NativePath file) @safe
+{
+	import std.array : appender;
+
+	// TODO: replace imagemagick command line tools with something like imageformats on dub
+
+	// use [0] to only get first frame in gifs, has no effect on static images.
+	string firstFrame = file.toNativeString ~ "[0]";
+
+	auto sizeInfo = execute(["identify", "-format", "%w %h %m", firstFrame]);
+	if (sizeInfo.status != 0)
+		return new LogoGenerateResponse(null, "Malformed image.");
```

Comment:
I would throw here, but it was broken when I tried

Reply:
Yeah I'm aware - `Future` doesn't really support exceptions like es6 promises.

Reply:
And it also can't, at least in a safe way, because of their thread-local nature in D, especially since the latest RC additions. It should require the callback to be nothrow, though.